lib/oscap.py

```diff
@@ -1,28 +1,144 @@
 import sys
 import re
+import enum
 import contextlib
+import collections
+import types
+import xml.etree.ElementTree as ET
 from pathlib import Path
 
 from lib import util, results
 
-_no_remediation_cache = None
 
-
-def _rules_without_remediation():
-    # TODO: parse this info from datastream XML
-    cmd = ['oscap', 'xccdf', 'generate', '--profile', '(all)', 'fix', util.get_datastream()]
-    _, lines = util.subprocess_stream(cmd, check=True)
-    for line in lines:
-        match = re.search('FIX FOR THIS RULE \'xccdf_org.ssgproject.content_rule_(.+)\' IS MISSING!', line)  # noqa
-        if match:
-            yield match.group(1)
-
-
-def has_no_remediation(rule):
-    global _no_remediation_cache
-    if _no_remediation_cache is None:
-        _no_remediation_cache = set(_rules_without_remediation())
-    return rule in _no_remediation_cache
+class Datastream:
+    FixType = enum.Flag('FixType', ['bash', 'ansible'])
+
+    def __init__(self, xml_file=None):
+        # extracted datastream metadata
+        #   self.profiles = {
+        #     'ospp': namespace(
+        #       .title = 'Some text',
+        #       .rules = set( 'audit_delete_success' , 'service_firewalld_enabled' , ...),
+        #       .values = set( ('var_rekey_limit_size','1G') , ...),
+        #     ),
+        #   }
+        #   self.rules = {
+        #     'configure_crypto_policy': namespace(
+        #       .fixes = FixType.bash | FixType.ansible,
+        #     ),
+        #     'account_password_selinux_faillock_dir': namespace(
+        #       .fixes = FixType.bash,
+        #     ),
+        #   }
+        def make_profile():
+            return types.SimpleNamespace(title=None, rules=set(), values=set())
+        self.profiles = collections.defaultdict(make_profile)
+        def make_rule():
+            return types.SimpleNamespace(fixes=self.FixType(0))
+        self.rules = collections.defaultdict(make_rule)
+        if xml_file:
+            self.parse_xml(xml_file)
+
+    def parse_xml(self, xml_file):
+        # parse input XML datastream in 10KB binary chunks (arbitrary reasonable value),
+        # pass them to the ElementTree parser, which returns element start/end events
+        parser = ET.XMLPullParser(events=['start', 'end'])
+        stack = []
+        with open(xml_file, 'rb') as f:
+            while True:
+                chunk = f.read(10000)
+                if not chunk:
+                    break
+                parser.feed(chunk)
+                for event, elem in parser.read_events():
+                    if event == 'start':
+                        stack.append(elem)
+                    else:
+                        # optimize a bit - filter out elements too shallow for anything below
+                        if len(stack) < 4:
+                            stack.pop()
+                            continue
+                        frames = [elem.tag.partition('}')[2] or elem.tag for elem in stack]
```

Comment:
Some comment here describing why we partition using '}' would help the readability, I guess it is because of namespaces right? E.g. `{http://scap.nist.gov/schema/scap/source/1.2}component-ref`.